# Task1

## Lady Tasting Tea

An event in Fisher's life led him to design this experiment, His friend's future wife, Muriel Bristol, said that she could tell if the tea or the milk was put first into a cup. She was given eight cups, four of them created each way and given to her in random order. Fisher wanted to find the probability that she could say the order of addition to the cup. In the end, Muriel correctly identified which of the four cups of tea had milk poured before the tea.

Problem Statement
Suppose the Lady Tasting Tea experiment is altered to involve twelve cups of tea. Six have the milk in first and the other six having tea in first. A person claims that they have the special power of being able to tell if the tea or the milk went into a cup first upon tasting it. You agree to accept their claim if they can tell which of the six cups in your experiment had the milk in first.
1.	Calculate, using Python, the probability that they select the correct six cups.
Suppose, now, you are willing to accept one error. Once they select the six cups they think had the milk in first, you will give them the benefit of the doubt should they have selected at least five of the correct cups.
2.	Calculate the probability, assuming they have no special powers, that the person makes at most one error.
3.	Would you accept two errors? Explain.



In [19]:
# Mathematical functions from the standard library.
# https://docs.python.org/3/library/math.html
import math
import itertools
import random
import numpy as np

For this task, we will be using 12 cups of tea as the test. 

In [2]:
# total number of cups.
cups = 12

# Milk first.
milk_first = 6

# Tea first.
tea_first = 6

We can show that from 12 cups, there is a possibility of choosing 1 from 12 choices, that leaves 11 cups to choose from. 1 out of 11 leaves 10 and so on.
So, there are 12 x 11 x 10 x 9 x 8 x 7 choices of cup as the test progresses. After 6 cups have been selected the rest are ignored. 

In [3]:
# Total ways of selecting 6 random cups.
selection = math.comb(cups, milk_first)

# Display total number of possible selections.
selection

924

Now that we know how many permutations there are of selecting 6 cups from 12, we want to see the number ways we can randomise the presentation of the cups to the tester

In [4]:
# How many chances are there that the lady will select correctly out of the 6 cups.
random = 6 * 5 * 4 * 3 * 2 * 1

# Display total number of possible selections.
random

720

In [18]:
labels = list(range(cups))

labels

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [21]:
user_combinations = list(itertools.combinations(labels, milk_first))

user_combinations

[(0, 1, 2, 3, 4, 5),
 (0, 1, 2, 3, 4, 6),
 (0, 1, 2, 3, 4, 7),
 (0, 1, 2, 3, 4, 8),
 (0, 1, 2, 3, 4, 9),
 (0, 1, 2, 3, 4, 10),
 (0, 1, 2, 3, 4, 11),
 (0, 1, 2, 3, 5, 6),
 (0, 1, 2, 3, 5, 7),
 (0, 1, 2, 3, 5, 8),
 (0, 1, 2, 3, 5, 9),
 (0, 1, 2, 3, 5, 10),
 (0, 1, 2, 3, 5, 11),
 (0, 1, 2, 3, 6, 7),
 (0, 1, 2, 3, 6, 8),
 (0, 1, 2, 3, 6, 9),
 (0, 1, 2, 3, 6, 10),
 (0, 1, 2, 3, 6, 11),
 (0, 1, 2, 3, 7, 8),
 (0, 1, 2, 3, 7, 9),
 (0, 1, 2, 3, 7, 10),
 (0, 1, 2, 3, 7, 11),
 (0, 1, 2, 3, 8, 9),
 (0, 1, 2, 3, 8, 10),
 (0, 1, 2, 3, 8, 11),
 (0, 1, 2, 3, 9, 10),
 (0, 1, 2, 3, 9, 11),
 (0, 1, 2, 3, 10, 11),
 (0, 1, 2, 4, 5, 6),
 (0, 1, 2, 4, 5, 7),
 (0, 1, 2, 4, 5, 8),
 (0, 1, 2, 4, 5, 9),
 (0, 1, 2, 4, 5, 10),
 (0, 1, 2, 4, 5, 11),
 (0, 1, 2, 4, 6, 7),
 (0, 1, 2, 4, 6, 8),
 (0, 1, 2, 4, 6, 9),
 (0, 1, 2, 4, 6, 10),
 (0, 1, 2, 4, 6, 11),
 (0, 1, 2, 4, 7, 8),
 (0, 1, 2, 4, 7, 9),
 (0, 1, 2, 4, 7, 10),
 (0, 1, 2, 4, 7, 11),
 (0, 1, 2, 4, 8, 9),
 (0, 1, 2, 4, 8, 10),
 (0, 1, 2, 4, 8, 11),
 (0, 1, 2, 4

In [5]:
# No of permutations that the tester can decide which liquid came first.
permutations = selection // random

# Show.
permutations

1

In [12]:
#The probability of selecting 6 cups corrctly from 12 is given below
probability = 1 / selection

percentage = probability * 100

print(f'The probability for the first 6 cups to be identifiedcorrectly is {probability} which is {percentage}%')

The probability for the first 6 cups being identified correctly is 0.0010822510822510823 or 0.1 %


Therefore, we can see that the chance of selecting the six cups with milk poured first from 12 cups is 1 in 924.
Or 1/924 which gives us a probability of 0.001 (rounded down) or 0.1%.

Calculate the probability of making at least one error

In [14]:
# Probability of selecting exactly 5 correct cups

five_right = math.comb(6, 5) * math.comb(6, 1)


probability_5 = five_right / selection

percentage_5 = probability_5 * 100
probability_5
print(f'The probability for the subject to identify 5 cups correctly is {probability_5} or {percentage:.1f} %')


The probability for the subject to identify 5 cups correctly is 0.03896103896103896 or 3.9 %


In [8]:
# Total probability of making at most one error
probability_1_wrong = probability_5 + permutations


probability_1_wrong

1.3222943722943723

To decide if we can accept two errors as reasonable, we have to calculate the probability of selecting 4 of the cups correctly out of 6:

In [17]:
four_right = math.comb(6, 4) * math.comb(6, 2)

probability_4 = four_right / selection

percentage_4 = probability_4 * 100

print(f'The probability for the subject to identify 4 cups correctly is {probability_4} or {percentage_4:.1f} %')

The probability for the subject to identify 4 cups correctly is 0.2435064935064935 or 24.4 %


In [10]:
probability_4 = four_right / selection

probability_4

0.2435064935064935

In conclusion, the lady has a 24.4 % chance to select the cups with 2 errors allowed, 3.9 % to select with 1 error and 0.1 % to select all 6 cups correctly out of 12.

We carry out the statistical tests to see if the claim can be verified. 
If not, we call it the Null Hypothesis, whereby we can not find any statistics to prove the claim is correct. In our experiment, the Null hypothesis (H₀) would indicate to us that Muriel cannot reliably discern if milk is put in first to the cup or not. We try to ensure she selects the six cups from the twelve cups at random.
Alternative hypothesis (H₁): She can tell the difference between cups with the milk in first or not.

### Task 2: Numpy's Normal Distribution

Task 2 is to assess whether `numpy.random.normal()` properly generates normal values.
To start, we generate a sample of one hundred thousand values using the function with mean `10.0` and standard deviation `3.0`.

Use the `scipy.stats.shapiro()` function to test whether the sample came from a normal distribution.
Explain the results and output.

Plot a histogram of the values and plot the corresponding normal distribution probability density function on top of it.

In [13]:
#
n = 100000
mu = 10.0
sigma = 3.0
n_bins = 30
# from lecture data = np.random.normal(size=100000)
#loc is mean of distrib, scale is the standard deviation so code is random.normal(loc=0.0, scale = 1.0, size = None)

sample = random.normal(loc=10.0, scale = 3.0, size = 100000)
#sample = np.random.default_rng(seed=100000)

AttributeError: 'int' object has no attribute 'normal'